#!/bin/sh /usr/share/dpatch/dpatch-run

## submenu patch
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Adds submenus within the commands and recording commands menu.
## DP: To create a submenu entry, prefix the name by one ore more "-".

@DPATCH@
diff -urNad vdr-1.3.23/config.c /tmp/dpep.74bcRU/vdr-1.3.23/config.c
--- vdr-1.3.23/config.c	2005-02-20 13:52:59.000000000 +0100
+++ /tmp/dpep.74bcRU/vdr-1.3.23/config.c	2005-04-09 21:28:28.912856112 +0200
@@ -27,18 +27,29 @@
 {
   title = command = NULL;
   confirm = false;
+  nIndent = 0;
+  childs = NULL;
 }
 
 cCommand::~cCommand()
 {
   free(title);
   free(command);
+  delete childs;
 }
 
 bool cCommand::Parse(const char *s)
 {
   const char *p = strchr(s, ':');
   if (p) {
+    nIndent = 0;
+#ifdef CMD_SUBMENUS
+    while (*s == '-')
+    {
+      nIndent++;
+      s++;
+    }
+#endif // CMD_SUBMENUS
      int l = p - s;
      if (l > 0) {
         title = MALLOC(char, l + 1);
@@ -85,6 +96,76 @@
   return result;
 }
 
+int cCommand::getIndent ()
+{
+  return nIndent;
+}
+
+void cCommand::setIndent (int nNewIndent)
+{
+  nIndent = nNewIndent;
+}
+
+bool cCommand::hasChilds ()
+{
+  if (!childs)
+  {
+    return false;
+  }
+  return (childs->Count () > 0);
+}
+
+int cCommand::getChildCount ()
+{
+  if (!childs)
+  {
+    return false;
+  }
+  return childs->Count ();
+}
+
+void cCommand::addChild (cCommand *newChild)
+{
+  if (!childs)
+  {
+    childs = new cCommands ();
+  }
+  childs->Add (newChild);
+}
+
+
+cCommands *cCommand::getChilds ()
+{
+  return childs;
+}
+
+// --- cCommands -------------------------------------------------------
+
+void cCommands::AddConfig(cCommand *Object)
+{
+  cCommand *c = (cCommand *) Object;
+  cCommand *cParent;
+  int nIndent;
+  int nIndex;
+
+  if (!c)
+  {
+    return;
+  }
+  nIndent = c->getIndent ();
+  //  isyslog ("nIndent %d %s\n", nIndent, c->Title ());
+  for (nIndex = Count () - 1; nIndex >= 0; nIndex--)
+  {
+    cParent = (cCommand *) Get (nIndex);
+    if (cParent->getIndent () < nIndent)
+    {
+      cParent->addChild (c);
+      return;
+    }
+  }
+  cConfig<cCommand>::Add(Object);
+}
+
 // -- cSVDRPhost -------------------------------------------------------------
 
 cSVDRPhost::cSVDRPhost(void)
diff -urNad vdr-1.3.23/config.h /tmp/dpep.74bcRU/vdr-1.3.23/config.h
--- vdr-1.3.23/config.h	2005-03-05 16:44:35.000000000 +0100
+++ /tmp/dpep.74bcRU/vdr-1.3.23/config.h	2005-04-09 21:28:28.913855960 +0200
@@ -35,11 +35,15 @@
 #define MaxSkinName 16
 #define MaxThemeName 16
 
+class cCommands;
+
 class cCommand : public cListObject {
 private:
   char *title;
   char *command;
   bool confirm;
+  int nIndent;
+  cCommands *childs;
   static char *result;
 public:
   cCommand(void);
@@ -48,6 +52,12 @@
   const char *Title(void) { return title; }
   bool Confirm(void) { return confirm; }
   const char *Execute(const char *Parameters = NULL);
+  int getIndent ();
+  void setIndent (int nNewIndent);
+  bool hasChilds ();
+  int getChildCount ();
+  cCommands *getChilds ();
+  void addChild (cCommand *newChild);
   };
 
 typedef uint32_t in_addr_t; //XXX from /usr/include/netinet/in.h (apparently this is not defined on systems with glibc < 2.2)
@@ -90,6 +100,10 @@
   cConfig(void) { fileName = NULL; }
   virtual ~cConfig() { free(fileName); }
   const char *FileName(void) { return fileName; }
+  virtual void AddConfig(T *Object)
+  {
+    cList<T>::Add(Object);
+  }
   bool Load(const char *FileName = NULL, bool AllowComments = false, bool MustExist = false)
   {
     Clear();
@@ -117,7 +131,7 @@
                 if (!isempty(buffer)) {
                    T *l = new T;
                    if (l->Parse(buffer))
-                      Add(l);
+                      AddConfig(l);
                    else {
                       esyslog("ERROR: error in %s, line %d\n", fileName, line);
                       delete l;
@@ -159,7 +173,10 @@
   }
   };
 
-class cCommands : public cConfig<cCommand> {};
+class cCommands : public cConfig<cCommand> {
+public:
+  virtual void AddConfig(cCommand *Object);
+  };
 
 class cSVDRPhosts : public cConfig<cSVDRPhost> {
 public:
diff -urNad vdr-1.3.23/Makefile /tmp/dpep.74bcRU/vdr-1.3.23/Makefile
--- vdr-1.3.23/Makefile	2005-04-09 21:28:28.756879824 +0200
+++ /tmp/dpep.74bcRU/vdr-1.3.23/Makefile	2005-04-09 21:28:28.912856112 +0200
@@ -67,6 +67,7 @@
 
 DEFINES += -DREMOTE_$(REMOTE)
 
+DEFINES += -DCMD_SUBMENUS
 DEFINES += -D_GNU_SOURCE
 
 DEFINES += -DVIDEODIR=\"$(VIDEODIR)\"
diff -urNad vdr-1.3.23/menu.c /tmp/dpep.74bcRU/vdr-1.3.23/menu.c
--- vdr-1.3.23/menu.c	2005-03-20 16:14:51.000000000 +0100
+++ /tmp/dpep.74bcRU/vdr-1.3.23/menu.c	2005-04-09 21:28:28.914855808 +0200
@@ -1272,6 +1272,12 @@
   if (command) {
      char *buffer = NULL;
      bool confirmed = true;
+#ifdef CMD_SUBMENUS
+     if (command->hasChilds()) {
+        AddSubMenu(new cMenuCommands(command->Title(), command->getChilds(), parameters));
+        return osContinue;
+        }
+#endif // CMD_SUBMENUS
      if (command->Confirm()) {
         asprintf(&buffer, "%s?", command->Title());
         confirmed = Interface->Confirm(buffer);
